### PR TITLE
ENH: constants: add array api support

### PR DESCRIPTION
**Pull request by j-bowhay**
#### Reference issue

towards #18867
#### What does this implement/fix?

Adds array API support to the three functions in `constants`
#### Additional information

Appreciate this is probably the least important module to add support for, however, it was an easy one to play around with. Happy to close if we don't want this!

### CODE REVIEW

**Comment by j-bowhay** [comment]
@lucascolley I was a bit confused about how to adapt the type hinting, any suggestions?

File: scipy/constants/_constants.py
Comment (lucascolley) [inline scipy/constants/_constants.py:272]: this line is unnecessary

File: scipy/constants/_constants.py
Comment (lucascolley) [inline scipy/constants/_constants.py:228]: Seems reasonable for now. We are waiting for progress on the array API end for type hints.

File: scipy/constants/_constants.py
Comment (lucascolley) [inline scipy/constants/_constants.py:334]: We have a float and an `xp` array. Should be either:

```suggestion
    return c / lambda_
```

or

```suggestion
    return xp.asarray(c) / lambda_
```

File: scipy/constants/_constants.py
Comment (lucascolley) [inline scipy/constants/_constants.py:365]: Same here

File: scipy/constants/_constants.py
Comment (j-bowhay) [reply scipy/constants/_constants.py:272]: Do you mind elaborating here for my understanding, do we not need this in case val is a list for example?

Comment (lucascolley) [reply scipy/constants/_constants.py:272]: right, ignore me, that seems correct

File: scipy/constants/_constants.py
Comment (j-bowhay) [reply scipy/constants/_constants.py:334]: I think these and the bellow might also been fine for the same reason as https://github.com/scipy/scipy/pull/20593#discussion_r1581759751

Comment (lucascolley) [reply scipy/constants/_constants.py:334]: Yes, as long as `a_float / an_xp_array == an_xp_array` in general.

**Review by lucascolley — CHANGES_REQUESTED** [review]
I think what you've done so far is good, but the tests need to check that the helper functions work with `xp` arrays. Currently, the `xp` parameter is just unused. The inputs should be `xp` arrays, and so should the expected values. That should confirm whether the correct namespaces are returned.

EDIT: you may want to keep a test / add cases for Python list input.

File: scipy/constants/_constants.py
Comment (mdhaber) [inline scipy/constants/_constants.py:283]: I didn't know about the equals sign in f-strings. That's convenient.

File: scipy/constants/_constants.py
Comment (mdhaber) [inline scipy/constants/_constants.py:274]: Will use of `asarray` instead of `asanyarray` break anything that is supposed to be supported?

**Review by mdhaber — COMMENTED** [review]
I thought I'd try to help here, but my main comment is the same as what @lucascolley wrote about the tests.

File: scipy/constants/_constants.py
Comment (j-bowhay) [reply scipy/constants/_constants.py:274]: According to https://stackoverflow.com/questions/59350359/any-examples-for-numpy-asanyarray-vs-asarray only if you are using subclasses of an `ndarray`, hopefully that is niche enough not to be an issue? This looks like a quirk of history rather than a change to in response to something https://github.com/scipy/scipy/commit/52f43e9f1b5391ddc7c30152318563c1f107a284

**Comment by j-bowhay** [comment]
Thanks both, I have updated the tests

**Comment by j-bowhay** [comment]
Looks like a tolerance bump might be required?

File: scipy/constants/_constants.py
Comment (lucascolley) [reply scipy/constants/_constants.py:274]: Maybe worth trying to find the niche case - the alternative would be `val = np.asanyarray(val) if is_numpy(xp) else xp.asarray(val)` I think. `array_namespace` returns a `numpy` namespace for anything unrecognised but coercible by `np.asanyarray` (that isn't otherwise refused).

Dropping it should probably be an explicit deprecation otherwise.

Comment (j-bowhay) [reply scipy/constants/_constants.py:274]: Or maybe we add a `subok` argument to `_asarray`?

Comment (mdhaber) [reply scipy/constants/_constants.py:274]: > Dropping it should probably be an explicit deprecation otherwise.

If it's advertised, sure, but I don't think it's needed if subclasses were supported by accident.

**Comment by mdhaber** [comment]
Re: tolerance bump, what do you think of changing the default of `xp_assert_close`?

https://github.com/scipy/scipy/blob/5e1cb66c827cb6debda2c28f09cefbca78ee5dd6/scipy/_lib/_array_api.py#L286-L299

It makes the default a little tighter for `float64` than it is now, and tests are passing.

File: scipy/constants/_constants.py
Comment (lucascolley) [reply scipy/constants/_constants.py:274]: It is advertised (as an `array_like`) - https://docs.scipy.org/doc/scipy/reference/generated/scipy.constants.convert_temperature.html#scipy.constants.convert_temperature

**Comment by lucascolley** [comment]
That sounds theoretically better, although I don't want to lose being able to substitute `xp_assert_allclose` with `np.testing.assert_allclose` 1-for-1. If there were lots of cases where we'd have to start adding tolerance arguments, that seems a bit messy. If not, then sounds good.

File: scipy/constants/_constants.py
Comment (mdhaber) [reply scipy/constants/_constants.py:274]: Then according to the [NumPy glossary](https://numpy.org/doc/stable/glossary.html), `{1: 0}` and `scipy` are valid inputs, too.

image

Comment (lucascolley) [reply scipy/constants/_constants.py:274]: 🙃 well that's a much broader question of how strictly do we have to stick by poor documentation. I don't know the answer.

Comment (j-bowhay) [reply scipy/constants/_constants.py:274]: Most of SciPy is advertised as `array like`, there are 2642 occurrences of `asarray` and 48 occurrences of `asanyarray` so I would be very surprised if anyone is relying on the behaviour of the slightly random selection of functions that use `asanyarray`. On the other hand, maybe I am being cavalier, happy to be convinced by other arguments!

Comment (lucascolley) [reply scipy/constants/_constants.py:274]: Right, in that case I agree that it isn't a big deal, but I think it would be worth it to officially 'fix' the documentation and deprecate the 48 occurrences of `asanyarray`. Thoughts?

Comment (mdhaber) [reply scipy/constants/_constants.py:274]: > On the other hand, maybe I am being cavalier, happy to be convinced by other arguments!

I don't think so. Starting this thread, I really just meant to check whether there were any tests or documentation that explicitly suggested that other arrays (e.g. masked arrays and matrices) should be accepted (as there are sometimes in stats). Since there aren't, I think it's safe enough.

Comment (mdhaber) [reply scipy/constants/_constants.py:274]: And yes, my point about `array_like` is that *technically* it's essentially meaningless. Pretty much the only thing that isn't allowed by `np.array` is a ragged array.
```python3
np.array([[1, 2, 3], [2, 3]])
```
And if we try to interpret it more strictly, it is reasonable to think that `np.array` *might actually by called* or that the object will be interpreted as a `np.ndarray` (not a subclass). So I think however you slice it, it's safe enough unless there's something more explicit in the documentation or tests.

**Comment by mdhaber** [comment]
> If there were lots of cases where we'd have to start adding tolerance arguments,

That's what I'm trying to avoid. We often have to special-case the tolerance argument for `torch` or ensure that `torch` is working with `float64` since it typically fails the default tolerance when it's working with its default `float32`.

**Comment by j-bowhay** [comment]
CI is green, I added float64 dtypes to keep pytorch happy without changing the preexisting tolerances. +1 for a more pytorch friendly default

**Comment by tupui** [comment]
Wait wasn't there some open PR to refactor the whole module? IIRC the main maintainer of this module (not a SciPy maintainer) has made a few PRs for that.

**Comment by tupui** [comment]
#20430 is the last one. But they have a few of these.

**Comment by j-bowhay** [comment]
> #20430 is the last one. But they have a few of these.

Yes, I have seen these but it is quite unclear which are ready for review. Many have major BC breaks too. I'm not sure they block this either as most seem to be to do with the constants themselves, not these functions

**Comment by tupui** [comment]
I don't know either, but I think it would be good to clarify. It would be a shame if we had to turn down some already done work because we would have this go in instead.

Though if this PR takes into account the new proposals then all good. I just wanted to be sure you all had that other person in mind.

File: scipy/constants/_constants.py
Comment (mdhaber) [reply scipy/constants/_constants.py:274]: I do think it's a good idea to change the documentation, maybe as we go through these, to specify:
- `array` (if that's how array-API compatible arrays are to be identified in documentation, following the examples of the array API standard docs, e.g. [`abs`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.abs.html))
- dimensionality
- dtype

For example, `n-D real floating array`. But there would need to be a standard language for that, and I don't know that one exists.

Comment (lucascolley) [reply scipy/constants/_constants.py:274]: Yep, I've written somewhere about this before but not sure where. There are some orthogonal issues:
- the 48 occurrences of `np.asanyarray` should probably be replaced with `np.asarray` unless there really are documented special cases. This may or may not require deprecations, depending on how much the behaviour can be deemed undocumented.
- parameters are variably described as `ndarray` or `array_like`, but in both cases it usually means 'anything that `np.asarray` can make sense of'. This should be replaced with something like `array` as you suggest, along with the other information. 
- We need to keep supporting array-likes with `np` arrays, but they will not be supported for other backends. Hence, there is even more reason to get away from saying `array_like`. We can document elsewhere that array-likes will still be supported with NumPy as the assumed backend.

File: scipy/constants/tests/test_constants.py
Comment (lucascolley) [inline scipy/constants/tests/test_constants.py:64]: can we add `np_only` tests for each of these tests, without the `xp.asarray`s? For the temperatures, can just make it one comparison. Of course, the implementation here does still work with the array-likes we want it to, but that relies on `array_namespace` behaving and being used correctly. It would be nice to have the test coverage (more or less) strictly increase.

**Review by lucascolley — CHANGES_REQUESTED** [review]
LGTM once my one comment about testing is addressed, and we have checked that we are happy with the `np.asanyarray` removal, and the static typing.

EDIT: and yes, this PR shouldn't interfere with the other constants PRs

**Comment by tylerjereddy** [comment]
fyi, I didn't see any additional failures when testing this branch with GPU backends (CuPy, torch).

File: scipy/constants/tests/test_constants.py
Comment (lucascolley) [inline scipy/constants/tests/test_constants.py:72]: Sorry, what I meant was this. With your current implementation, `python dev.py test -b numpy -b pytorch -b cupy` would run the exact same assertion 3 times, which seems a bit wasteful.

```suggestion

@skip_xp_backends(np_only=True, reasons=['Python list input uses NumPy backend'])
def test_nu_to_lambda_array_like(xp):
    xp_assert_equal(sc.nu2lambda([sc.speed_of_light, 1]),
                    [1, sc.speed_of_light])
```

File: scipy/constants/_constants.py
Comment (rgommers) [reply scipy/constants/_constants.py:274]: Ndarray subclass support is not in great/consistent shape, but I also don't see the need to break it if a one-liner keeps things unchanged.

This is covered in gh-18286:

| Input type                | Return type          | When                          | Behavior                  | Notes                                                                               |
|---------------------------|----------------------|-------------------------------|---------------------------|-------------------------------------------------------------------------------------|
| numpy.ndarray subclasses  | same subclass type   | always                        |                           | so always use `np.asanyarray` checks                                                |

File: scipy/constants/_constants.py
Comment (rgommers) [reply scipy/constants/_constants.py:228]: I would just not touch type annotations. Doing so is both unnecessary as long as array API support is experimental, and a regressions for current type annotation users. It'd be better to deal with all type annotations for array-like arguments in one go.

File: scipy/constants/_constants.py
Comment (rgommers) [inline scipy/constants/_constants.py:312]: This is still array-like

**Review by rgommers — COMMENTED** [review]
> it was an easy one to play around with. Happy to close if we don't want this!

Thanks @j-bowhay! I think we do want this PR, as it can be useful and it's a small diff to add array API support.

My main comment is to keep the diff as small as possible, and keep it as focused on _adding_ array API support and tests. In particular:
- don't change return types for `numpy.ndarray` subclasses
- don't change type annotations
- don't remove `array_like` from docs

> EDIT: and yes, this PR shouldn't interfere with the other constants PRs

Agreed with comments on these. It's good to think about it, however this diff is small enough that it should not impact other long-running PRs.

File: scipy/constants/_constants.py
Comment (j-bowhay) [reply scipy/constants/_constants.py:228]: Happy to leave alone but wasn't sure how to keep mypy happy in the meantime?

Comment (rgommers) [reply scipy/constants/_constants.py:228]: Was it actually complaining? We don't run Mypy in the array-api CI job right?

Comment (j-bowhay) [reply scipy/constants/_constants.py:228]: Originally I did not change the type hints, see https://github.com/scipy/scipy/actions/runs/8858072080/job/24326283132
image

File: scipy/constants/tests/test_constants.py
Comment (j-bowhay) [inline scipy/constants/tests/test_constants.py:12]: I already had to touch all lines here so grouping the tests does not really increase the diff

File: scipy/constants/_constants.py
Comment (rgommers) [reply scipy/constants/_constants.py:230]: Hmm okay, not immediately clear to me what is going on there. Let me have a closer look. This will be a problem in other places where we have type annotations as well then I suspect.

File: scipy/constants/_constants.py
Comment (lucascolley) [reply scipy/constants/_constants.py:274]: How about following Jake's suggestion with `_asarray(x, ..., xp=None, ..., sub_ok=True)` which uses `np.asanyarray` instead of `np.asarray` for NumPy, but still `xp.asarray` for alternative backends?

File: scipy/constants/tests/test_constants.py
Comment (lucascolley) [inline scipy/constants/tests/test_constants.py:60]: This can also be `np_only`

File: scipy/constants/_constants.py
Comment (rgommers) [reply scipy/constants/_constants.py:274]: That sounds good to me.

File: scipy/constants/_constants.py
Comment (lucascolley) [inline scipy/constants/_constants.py:334]:
```suggestion
    return c / _asarray(lambda_, xp=xp, subok=True)
```

**Review by lucascolley — APPROVED** [review]
LGTM once the static typing is resolved!

File: scipy/constants/_constants.py
Comment (rgommers) [reply scipy/constants/_constants.py]: The problem turned out to be fairly straightforward to resolve once I figured out what it was: https://mypy.readthedocs.io/en/stable/common_issues.html#redefinitions-with-incompatible-types. When you do something like:
```
val = _asarray(val, xp=xp, subok=True)
```
and the type of `val` changes in any way other than it becoming more narrow, Mypy gets really unhappy. All the errors in the screenshot above are for Mypy reusing `npt.Arraylike` instead of `np.ndarray` after the change to add array API support. `ndarray` was narrower, `Any` is not.

Unfortunately Mypy gives extremely confusing errors for this, rather than just telling you "hey, this type isn't more narrow, I am reusing the originally declared one".

I pushed a fix that should resolve it. 

Once this is merged, I have type annotations for `_lib/_array_api.py` ready to go, but they're not actually needed for this PR.

**Comment by j-bowhay** [comment]
Thanks for the fix @rgommers I think this is good from my end now

**Review by rgommers — APPROVED** [review]
Okay, everyone and CI are happy - let's give this a go:) Thanks @j-bowhay & reviewers!

**Pull request merged**
